browse/controllers/abs_page/__init__.py

```diff
@@ -189,50 +187,51 @@ def _check_request_headers(docmeta: DocMetadata,
-    # them to the last modified time to determine whether we will return a
-    # "not modified" response
-    mod_since_dt = _time_header_parse(headers, 'If-Modified-Since')
-    none_match_dt = _time_header_parse(headers, 'If-None-Match')
-    not_modified = _not_modified(last_mod_dt, mod_since_dt, none_match_dt)
-
     last_mod_mime = mime_header_date(last_mod_dt)
-    headers['Last-Modified'] = last_mod_mime
-    headers['ETag'] = last_mod_mime
-    headers['Expires'] = abs_expires_header()[1]
+    etag = f'"{last_mod_mime}"'
+    
+    resp_headers['Last-Modified'] = last_mod_mime
+    resp_headers['ETag'] = etag
+    resp_headers['Expires'] = abs_expires_header()[1]
+    
+    not_modified = _not_modified(last_mod_dt,
+                                 _time_header_parse('If-Modified-Since'),
+                                 _get_req_header('if-none-match'),
+                                 etag)
 
     return not_modified
 
 
 def _not_modified(last_mod_dt: datetime,
                   mod_since_dt: Optional[datetime],
-                  none_match_dt: Optional[datetime])->bool:
-    if mod_since_dt and none_match_dt:
-        not_modified = (mod_since_dt >= last_mod_dt
-                        and none_match_dt >= last_mod_dt)
-    elif mod_since_dt and not none_match_dt:
-        not_modified = mod_since_dt >= last_mod_dt
-    elif none_match_dt and not mod_since_dt:
-        not_modified = none_match_dt >= last_mod_dt
+                  none_match: Optional[str],
+                  current_etag: str) -> bool:
+    if none_match and none_match == current_etag:
+        return True
+    elif mod_since_dt:
+        return (mod_since_dt.replace(microsecond=0) >=
+                last_mod_dt.replace(microsecond=0))
     else:
-        not_modified = False
-    return not_modified
+        return False
 
 
-def _time_header_parse(headers: Dict[str, Any], header: str) \
-        -> Optional[datetime]:
-    if (header in request.headers
-            and request.headers[header] is not None):
-        try:
-            dt = parser.parse(str(request.headers.get(header)))
-            if not dt.tzinfo:
-                dt = dt.replace(tzinfo=tzutc())
-            return dt
-        except (ValueError, TypeError):
-            print(f'Exception parsing the If-None-Match request header')
-            return None
-    else:
-        return None
+def _time_header_parse(header: str) -> Optional[datetime]:
+    try:
+        dt = parser.parse(str(_get_req_header(header)))
+        if not dt.tzinfo:
+            dt = dt.replace(tzinfo=tzutc())
+        return dt
+    except (ValueError, TypeError, KeyError):
+        pass
+    return None
+
+
+def _get_req_header(header: str) -> Optional[str]:
+    """Gets request header, needs to be case insensative for keys.
+
+    HTTP header keys are case insensative. RFC 2616"""
+    return next((value for key, value in request.headers.items()
```

Comment:
The HTTP header keys need to be accessed in a case insensative way. So if the client can send If-None-Match or if-none-match or If-NoNe-mAtCh.